server/async_tcp.go

```diff
@@ -142,9 +181,21 @@ func (s *AsyncServer) Run(ctx context.Context) error {
 		return err
 	}
 
-	err = s.eventLoop(ctx)
+	eventLoopCtx, cancelEventLoop := context.WithCancel(ctx)
+	defer cancelEventLoop()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		err = s.eventLoop(eventLoopCtx)
+		if err != nil {
+			cancelWatch()
```

Comment:
I feel there are two advantages if we log and return from here.
1. The differs would run as the execution flow exits the function.
2. The control would in main.go would invoke the common InitiateShutdown i mentioned earlier.

Reply:
Please ignore these comments. My bad...